src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+        return Token.create version, Array.chooseV tryDecode events }
+    member _.LoadFromToken(useWriteConn, streamName, log, token, tryDecode)
+        : Async<StreamToken * 'event array> = async {
+        let streamPosition = token.version + 1L
+        let connToUse = if useWriteConn then connection.WriteConnection else connection.ReadConnection
+        let! version, events = Read.loadForwardsFrom log connection.ReadRetryPolicy connToUse batchOptions.BatchSize batchOptions.MaxBatches streamName streamPosition
+        return Token.create (max token.version version), Array.chooseV tryDecode events }
+
+    member _.TrySync(log, streamName, token, encodedEvents : IEventData<EventBody> array): Async<GatewaySyncResult> = async {
+        match! Write.writeEvents log connection.WriteRetryPolicy connection.WriteConnection streamName token.version encodedEvents with
+        | MdbSyncResult.ConflictUnknown ->
+            return GatewaySyncResult.ConflictUnknown
+        | MdbSyncResult.Written version' ->
+            let token = Token.create version'
+            return GatewaySyncResult.Written token }
+
+    member _.Sync(log, streamName, streamVersion, events : IEventData<EventBody> array) : Async<GatewaySyncResult> = async {
+        let encodedEvents : IEventData<EventBody> array = events
+        match! Write.writeEvents log connection.WriteRetryPolicy connection.WriteConnection streamName streamVersion encodedEvents with
+        | MdbSyncResult.ConflictUnknown ->
+            return GatewaySyncResult.ConflictUnknown
+        | MdbSyncResult.Written version' ->
+            let token = Token.create version'
+            return GatewaySyncResult.Written token }
+
+[<NoComparison; NoEquality; RequireQualifiedAccess>]
+type AccessStrategy =
+    /// Load only the single most recent event defined in in a stream and trust that it'll be decoded and
+    /// doing a <c>fold</c> from any such event will yield a correct and complete state
+    /// In other words, the <c>fold</c> function should not need to consider either the preceding <c>'state</state> or <c>'event</c>s.
+    | LatestKnownEvent
+type private Category<'event, 'state, 'context>(context : MessageDbContext, codec : IEventCodec<_, _, 'context>, ?access) =
+    let tryDecode = codec.TryDecode
+    let loadAlgorithm load streamName log =
+        match access with
+        | None -> load (context.LoadBatched(streamName, log, tryDecode))
+        | Some AccessStrategy.LatestKnownEvent -> load (context.LoadLast(streamName, log, tryDecode))
+    let load (fold : 'state -> 'event seq -> 'state) initial f = async {
+        let! token, events = f
+        return struct (token, fold initial events) }
+    member _.Load(fold : 'state -> 'event seq -> 'state) (initial : 'state) (streamName : string) (log : ILogger) : Async<struct (StreamToken * 'state)> =
+        loadAlgorithm (load fold initial) streamName log
```

Comment:
if `load` was not passed in, this could be something like
```suggestion
        (load fold initial) (loadAlgorithm streamName log)
```

Reply:
Not sure why, but `(f a b) (g c d)` feels very hard to grok

Reply:
at least on a skim

Reply:
have done it anyway cause I agree load shouldn't be passed